Wip confidence levels

JamesSG2 opened the pull request:
- Added a  'low_confidence_rejection'  function to the backend.py (with comments) which combined with other changes I made, finds the confidence level for each prediction, and removes the words and their respective letters that are below the minimum confidence Level (currently 55% confidence -- is adjustable). 
- Can discuss further in the next meeting why a CL works for words but can't work on the character by character basis. 
- **Currently this system rejects about 50% of all incorrect words, in exchange for rejecting ~5% of correct words.** 
- I also spaced out the "Analyze" results to make them easier to read. 
- A system for alerting the user if all instances of a letter were rejected (or never existed) should be the next step. 
- Also no need to delete Wip-confidence-levels branch after merging; I can rename it and use for my next project / feature

JamesSG2 committed:
Confidence-Level-wip

Will discuss further in next meeting. Working getting confidence levels involved with Trevor's code (so fewer errors make it into the output char folder). Also tested new handwriting data (my brother's).

JamesSG2 committed:
Updating wip-C-L

Updating wip-C-L with newest code from dev branch.

JamesSG2 committed:
Merge branch 'dev' into wip-confidence-levels

JamesSG2 committed:
Update README.md

JamesSG2 committed:
Merge branch 'dev' into wip-confidence-levels

JamesSG2 committed:
Added CL Rejection Mechanism

Finds the confidence level for each prediction, and removes words and their respective letters below the minimum confidence Level (currently 55% confidence). System for alerting user if all instances of a letter were rejected (or never existed) should be the next step.

JamesSG2 committed:
Fixed minor spelling error

Also spaced out the "Analyze" results to make them easier to read (on last commit).

JamesSG2 committed:
For dev compatability

rover-t commented:
Looks good. Definitely an improvement over what it was before. The only issue I ran into is that I had to `pip install pandas` to get line 35 `output_type='data.frame'` to work. I would add that to the `requirements.txt` file so that it is listed on the requirements page. 

I also noticed that the `low_confidence_rejection` method directly acts on `pytesseract_list` rather than returning the corrected list separately. This works, but it might be better to return an updated list rather than acting on an existing list. I'm not really sure though, so I'll leave it up to you if you think we should change it.

Also worth playing around with - is there a way to run the regex correction I wrote before running your function? The more original text there is the more effective the regex will be. If that's not possible it seems great as is, but its just something I thought of.

JamesSG2 committed:
Made Suggested Changes

As suggested by Trevor:
- Updated requirements.txt (I already had pandas installed which is why I didn't initially realize it was required)
- LCR function now returns the  'confident_pytesseract_list' and doesn't edit 'pytesseract_list'
- I'm not familiar with the regex correction or if has a confidence level mechanism. It might cut down on the # of rejections though, which is good. It would require a completely different function though to run a CL rejection after the regex correction (unless someone changes it to edit pytesseract_list instead)

JamesSG2 commented:
I'll go ahead and merge wip_confidence_levels to dev now since it's been reviewed and I just made the suggested changes.
